containers/src/Data/IntMap/Internal.hs

```diff
@@ -2614,6 +2618,101 @@ partitionWithKey predicate0 t0 = toPair $ go predicate0 t0
           | otherwise     -> (Nil :*: t)
         Nil -> (Nil :*: Nil)
 
+-- | \(O(\min(n,W))\). Take while a predicate on the keys holds.
+-- The user is responsible for ensuring that for all @Int@s, @j \< k ==\> p j \>= p k@.
+-- See note at 'spanAntitone'.
+--
+-- @
+-- takeWhileAntitone p = 'fromDistinctAscList' . 'Data.List.takeWhile' (p . fst) . 'toList'
+-- takeWhileAntitone p = 'filterWithKey' (\\k _ -> p k)
+-- @
+--
+-- @since FIXME
+takeWhileAntitone :: (Key -> Bool) -> IntMap a -> IntMap a
+takeWhileAntitone predicate t =
+  case t of
+    Bin _ m l r
+      | m < 0 ->
+        if predicate 0 -- handle negative numbers.
+        then union r (go predicate l)
```

Comment:
Surely we have something faster than `union` for this?

Reply:
Ah thanks for catching this. I adapted these functions from `split` and didn't think too much about `union` since it will run in O(1) anyway, but we should just be using `bin`.

We should also update `split` and `splitMember` to use `bin`, but perhaps that should be a separate PR.

Reply:
Yeah, I wouldn't think it would be a big diff, but the simplest function that can get the job done will be best. Updating the splits should be a separate commit, ideally; it's up to you whether to make a separate PR.

Reply:
I can see a couple of other improvements for `split` and `splitMember` (adding benchmarks, strictness) so I think I'll do that separately.